### PR TITLE
GUI: Implement MSELF extraction tool

### LINKED ISSUE #9681: [Feature Request] GUI: Implement MSELF decryption in decryption tool
Can use names of file records in extracted files.

### PR DESCRIPTION
**Pull request by elad335**
Fixes #9681

### CODE REVIEW

File: rpcs3/rpcs3qt/main_window.cpp [hunk 1]
```diff
@@ -769,6 +771,88 @@ void main_window::HandlePackageInstallation(QStringList file_paths)
...
+{
+	const QString path_last_mself = m_gui_settings->GetValue(gui::fd_extract_mself).toString();
+	QString file_path = QFileDialog::getOpenFileName(this, tr("Select MSELF To extract"), path_last_mself, tr("All pup files (*.mself);;All files (*.*)"));
+
+	QString dir = file_path.isEmpty() ? QString{} : QFileDialog::getExistingDirectory(this, tr("Extraction Directory"), QString{}, QFileDialog::ShowDirsOnly | QFileDialog::DontResolveSymlinks);
+
+	if (!dir.isEmpty())
+	{
+		HandleMSELFextraction(file_path, dir);
+	}
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp:784]:
```suggestion
	if (!file_path.isEmpty())
	{
		const QString dir = QFileDialog::getExistingDirectory(this, tr("Extraction Directory"), QString{}, QFileDialog::ShowDirsOnly | QFileDialog::DontResolveSymlinks);

		if (!dir.isEmpty())
		{
			HandleMSELFextraction(file_path, dir);
		}
	}
```

File: rpcs3/rpcs3qt/main_window.cpp [hunk 2]
```diff
@@ -769,6 +771,88 @@ void main_window::HandlePackageInstallation(QStringList file_paths)
...
+}
+
+void main_window::HandleMSELFextraction(QString file_path, QString dir_path)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp:787]: why is this part of the main window?
There is no reason to implement this in a gui dialog

Comment (elad335) [reply rpcs3/rpcs3qt/main_window.cpp:788]: Where'd you put it?

Comment (Megamouse) [reply rpcs3/rpcs3qt/main_window.cpp:788]: mself obviously

**Comment by elad335** [comment]
Added mself.cpp

File: rpcs3/Emu/CMakeLists.txt [hunk 3]
```diff
@@ -102,6 +102,7 @@ target_sources(rpcs3_emu PRIVATE
 	../Loader/PSF.cpp
 	../Loader/PUP.cpp
 	../Loader/TAR.cpp
+	../Loader/MSELF.cpp
```
Comment (Megamouse) [inline rpcs3/Emu/CMakeLists.txt:105]: So far these entries were sorted

Comment (elad335) [reply rpcs3/Emu/CMakeLists.txt:105]: Fixed.

File: rpcs3/Loader/mself.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+		mself.seek(pos);
+		mself.read(buffer.data(), rec.size);
+
+		if (!fs::write_file(extract_to + name, fs::rewrite, buffer))
+		{
+			mself_log.error("Error creating %s (%s)", extract_to + name, fs::g_tls_error);
+			return false;
+		}
+
+		mself_log.success("Extraction of %s from MSELF ccompleted. (its path: '%s')", name, extract_to + name);
```
Comment (Megamouse) [inline rpcs3/Loader/mself.cpp:72]: The log channel already tells us that this is about MSELF. Can probably be simplified in other log messages as well
```suggestion
		mself_log.success("Extracted %s to '%s'", name, extract_to + name);
```

File: rpcs3/Loader/mself.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	mself_log.success("MSELF extraction complete!");
```
Comment (Megamouse) [inline rpcs3/Loader/mself.cpp:75]:
```suggestion
	mself_log.success("Extraction complete!");
```

File: rpcs3/Loader/mself.hpp [hunk 6]
```diff
@@ -12,10 +12,10 @@ struct mself_header
 	be_t<u32> header_size; // ???
 	u8 reserved[0x28];
 
-	u32 get_count(u64 file_size)
+	u32 get_count(u64 file_size) const
 	{
 		// Fast sanity check
-		if (magic != "MSF"_u32 || ver != u32{1} || this->size != file_size) [[unlikely]]
+		if (magic != "MSF"_u32 || ver != u32{1} || (file_size - sizeof(mself_header)) / 0x40 < count || this->size != file_size) [[unlikely]]
```
Comment (Megamouse) [inline rpcs3/Loader/mself.hpp:18]: Please replace both occasions of 0x40 with some constant.
It's hard to tell what's the logic here if you use a "random" number

**Comment by Megamouse** [comment]
I made a pull request on this branch with Visual Studio fixes

File: rpcs3/Loader/mself.cpp (hunk 5 above)
Comment (elad335) [reply rpcs3/Loader/mself.cpp]: fixed.

File: rpcs3/Loader/mself.hpp (hunk 6 above)
Comment (elad335) [reply rpcs3/Loader/mself.hpp]: fixed.

File: rpcs3/Loader/mself.hpp [hunk 7]
```diff
@@ -12,10 +29,10 @@ struct mself_header
 	be_t<u32> header_size; // ???
 	u8 reserved[0x28];
 
-	u32 get_count(u64 file_size)
+	u32 get_count(u64 file_size) const
 	{
 		// Fast sanity check
-		if (magic != "MSF"_u32 || ver != u32{1} || this->size != file_size) [[unlikely]]
+		if (magic != "MSF"_u32 || ver != u32{1} || (file_size - sizeof(mself_header)) / sizeof(mself_record) < count || this->size != file_size) [[unlikely]]
```
Comment (Megamouse) [inline rpcs3/Loader/mself.hpp:35]: does this handle overflow if file_size < 0x40 ?

Comment (elad335) [reply rpcs3/Loader/mself.hpp:35]: It must be over 0x40 because otherwise the file header could not be read.

**Pull request merged**